File: integrations/browser/oicana-browser/src/CompiledDocument.ts
import {
  document_pages,
  export_document,
  remove_document,
} from '@oicana/browser-wasm';
import { type ExportFormat, Png } from './ExportFormat.js';
import type { PageRange } from './PageRange.js';

/**
 * Size of a single document page, in typographic points (pt).
 */
export interface PageSize {
  /** Page width in points. */
  width: number;
  /** Page height in points. */
  height: number;
}

/**
 * A compiled document. Its pages can be exported on
 * demand without re-compiling.
 *
 * Obtain one via {@link Template.compile}. Call {@link dispose} (or
 * use the `using` keyword) to release the underlying document.
 */
export class CompiledDocument implements Disposable {
  private documentId: string | undefined;

  /** Sizes (in points) of every page, in document order. */
  public readonly pages: ReadonlyArray<PageSize>;

  /**
   * @internal Construct via {@link Template.compile}.
   */
  public constructor(documentId: string) {
    this.documentId = documentId;
    this.pages = JSON.parse(document_pages(documentId)) as PageSize[];
  }

  /** Number of pages in the document. */
  public get pageCount(): number {
    return this.pages.length;
  }

  /**
   * Export a single page of the document to PNG.
   * @param pageIndex - zero-based index of the page to export
   * @param pixelsPerPt - resolution in pixels per point
   */
  public exportPage(pageIndex: number, pixelsPerPt: number): Uint8Array {
    return this.export(Png(pixelsPerPt), {
      start: pageIndex + 1,
      end: pageIndex + 1,
    });
  }

  /**
   * Export the document in the given format (defaults to PDF), optionally
   * restricted to a range of pages.
   * @param format - export format specification
   * @param pages - 1-based, inclusive page range (defaults to the whole document)
   */
  public export(
    format: ExportFormat = { format: 'pdf' },
    pages?: PageRange,
  ): Uint8Array {
    if (this.documentId === undefined) {
      throw new Error('CompiledDocument has already been disposed');
    }
    return export_document(this.documentId, format, pages);
  }

  /**
   * Export the document to a PDF file, optionally restricted to a range of pages.
   * @param pages - 1-based, inclusive page range (defaults to the whole document)
   */
  public toPdf(pages?: PageRange): Uint8Array {
    return this.export({ format: 'pdf' }, pages);
  }

  /**
   * Release the cached document. After calling dispose() this instance must not
   * be used anymore.
   */
  public dispose(): void {
    if (this.documentId !== undefined) {
      remove_document(this.documentId);
      this.documentId = undefined;
    }
  }

  /**
   * Enables use with the `using` keyword for automatic resource cleanup.
   */
  [Symbol.dispose](): void {
    this.dispose();
  }
}

File: integrations/csharp/Oicana/CompiledDocument.cs
using System.Text.Json.Nodes;
using Oicana.Interop;
using ExportFormat = Oicana.Config.ExportFormat;
using PageRange = Oicana.Config.PageRange;

namespace Oicana;

/// <summary>
/// A compiled document kept in memory so it can be exported on demand without re-compiling.
///
/// Obtain one via <see cref="Template.Compile"/>. Call <see cref="Dispose"/> (or use a
/// <c>using</c> statement) to free the underlying document.
/// </summary>
public sealed class CompiledDocument : IDisposable
{
    private string? _documentId;

    /// <summary>
    /// Sizes (in points) of every page, in document order.
    /// </summary>
    public IReadOnlyList<PageSize> Pages { get; }

    /// <summary>
    /// Warnings produced by the compilation of this document, or <c>null</c> if there were none.
    /// </summary>
    public string? Warnings { get; }

    /// <summary>
    /// Construct a handle around an already-compiled document. Use <see cref="Template.Compile"/>.
    /// </summary>
    /// <param name="documentId">Identifier of the compiled document in the internal cache.</param>
    internal CompiledDocument(string documentId)
    {
        _documentId = documentId;
        Pages = ParsePageSizes(OicanaFfi.DocumentPages(documentId));
        Warnings = OicanaFfi.GetWarnings(documentId);
    }

    /// <summary>
    /// Number of pages in the document.
    /// </summary>
    public int PageCount => Pages.Count;

    /// <summary>
    /// Export the document in the given format, optionally restricted to a range of pages.
    /// </summary>
    /// <param name="exportFormat">Format configuration for the document export.</param>
    /// <param name="pages">0-based, inclusive page range to export, or <c>null</c> for the whole document.</param>
    /// <exception cref="OicanaException">If the export fails.</exception>
    /// <returns>Stream containing the exported document.</returns>
    public Stream Export(ExportFormat exportFormat, PageRange? pages = null)
    {
        EnsureOpen();
        return OicanaFfi.ExportDocument(_documentId!, exportFormat, pages);
    }

    /// <summary>
    /// Export the document to PDF, optionally restricted to a range of pages.
    /// </summary>
    /// <param name="pages">0-based, inclusive page range to export, or <c>null</c> for the whole document.</param>
    /// <exception cref="OicanaException">If the export fails.</exception>
    /// <returns>Stream containing the exported PDF.</returns>
    public Stream ExportPdf(PageRange? pages = null)
    {
        return Export(ExportFormat.Pdf(), pages);
    }

    /// <summary>
    /// Export the document to PNG, optionally restricted to a range of pages.
    /// </summary>
    /// <param name="pixelsPerPt">Resolution in pixels per point (defaults to 1.0).</param>
    /// <param name="pages">0-based, inclusive page range to export, or <c>null</c> for the whole document.</param>
    /// <exception cref="OicanaException">If the export fails.</exception>
    /// <returns>Stream containing the exported PNG.</returns>
    public Stream ExportPng(float pixelsPerPt = 1.0f, PageRange? pages = null)
    {
        return Export(ExportFormat.Png(pixelsPerPt), pages);
    }

    /// <summary>
    /// Export the document to SVG, optionally restricted to a range of pages.
    /// </summary>
    /// <param name="pages">0-based, inclusive page range to export, or <c>null</c> for the whole document.</param>
    /// <exception cref="OicanaException">If the export fails.</exception>
    /// <returns>Stream containing the exported SVG.</returns>
    public Stream ExportSvg(PageRange? pages = null)
    {
        return Export(ExportFormat.Svg(), pages);
    }

    /// <summary>
    /// Release the cached document. The instance must not be used after disposal.
    /// </summary>
    public void Dispose()
    {
        if (_documentId != null)
        {
            OicanaFfi.RemoveDocument(_documentId);
            _documentId = null;
        }
    }

    private void EnsureOpen()
    {
        if (_documentId == null)
        {
            throw new ObjectDisposedException(nameof(CompiledDocument));
        }
    }

    private static IReadOnlyList<PageSize> ParsePageSizes(string json)
    {
        var array = JsonNode.Parse(json)?.AsArray() ?? new JsonArray();
        var pages = new List<PageSize>(array.Count);
        foreach (var node in array)
        {
            pages.Add(new PageSize(
                node!["width"]!.GetValue<double>(),
                node["height"]!.GetValue<double>()));
        }

        return pages;
    }
}
